File: src/engine/actor.lua
---@class Actor
---@field spriteEngine SpriteEngine
---@field x number
---@field y number
---@field speed number
---@field sprite Sprite
---@field parent Actor|nil
Actor = {}
Actor.__index = Actor

---@param spriteEngine SpriteEngine
---@param x number
---@param y number
---@param speed number
---@param sprite Sprite|nil
---@param parent Actor|nil
function Actor:new(spriteEngine, x, y, speed, sprite, parent)
    local obj = {
        spriteEngine = spriteEngine,
        x = x,
        y = y,
        speed = speed or 200,
        sprite = sprite,
        parent = parent
    }
    setmetatable(obj, Actor)
    return obj
end

---@param self Actor
---@param dt number
function Actor:update(dt)
    self.sprite:update(dt)
end

---@return number
function Actor:globalX()
    return self.x + (self.parent and self.parent.x or 0)
end
---@return number
function Actor:globalY()
    return self.y  + (self.parent and self.parent.y or 0)
end


-- Draw method
function Actor:draw()
    self.sprite:draw(self:globalX(), self:globalY())
end


-- function Actor:translate()
--     love.graphics.push()
--     love.graphics.translate(self.x, self.y)
-- end

-- function Actor:endTranslate()
--     love.graphics.pop()
-- end

return Actor

File: src/engine/actor.ts
import { Screen } from "love.graphics";
import { Collider2d } from "./collision/collider2d";
import { SpriteEngine } from "./sprite-engine";
import { Sprite } from "./sprite/sprite";
import { Vector2 } from "./tools";

export interface Damageable {
  isInvincible(): boolean;
  damage(src: Actor | undefined, amount: number): void;
}

export const idGenerator = {
  currentValue: 0,
  next() {
    return this.currentValue++;
  },
};

export class Actor {
  constructor(
    public id: number,
    private type: string,
    public spriteEngine: SpriteEngine,
    public pos: Vector2 = new Vector2(0, 0),
    public speed = 200,
    public sprite?: Sprite,
    public collider?: Collider2d,
    public parent?: Actor
  ) {
    if (collider) {
      collider.setParent(this);
    }
  }

  getType(): string {
    return this.type;
  }

  update(dt: number) {
    this.sprite?.update(dt);
  }

  globalPos(): Vector2 {
    return this.parent ? this.pos.add(this.parent.pos) : this.pos;
  }

  draw(screen?: Screen) {
    this.sprite?.draw(this.pos, screen);
  }

  serialize?(): string;
  deserialize?(data: string): void;
}
